base/src/main/java/io/spine/validate/ConstraintInterpreter.java

```diff
@@ -0,0 +1,300 @@
+/*
+ * Copyright 2019, TeamDev. All rights reserved.
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.validate;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Range;
+import com.google.protobuf.Message;
+import io.spine.base.FieldPath;
+import io.spine.code.proto.FieldContext;
+import io.spine.code.proto.FieldDeclaration;
+import io.spine.code.proto.FieldName;
+import io.spine.option.PatternOption;
+import io.spine.type.MessageType;
+import io.spine.type.TypeName;
+import io.spine.validate.option.DistinctConstraint;
+import io.spine.validate.option.GoesConstraint;
+import io.spine.validate.option.PatternConstraint;
+import io.spine.validate.option.RangedConstraint;
+import io.spine.validate.option.RequiredConstraint;
+import io.spine.validate.option.RequiredFieldConstraint;
+import io.spine.validate.option.ValidateConstraint;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.spine.protobuf.AnyPacker.unpackIfPacked;
+import static io.spine.protobuf.TypeConverter.toAny;
+import static io.spine.util.Exceptions.newIllegalStateException;
+import static io.spine.validate.MessageValue.atTopLevel;
+import static io.spine.validate.MessageValue.nestedIn;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Interprets validation constraints by applying them to a given message value.
+ *
+ * <p>The output result of this {@link ConstraintTranslator} is a {@link ValidationError}.
+ */
+@SuppressWarnings("OverlyCoupledClass")
+final class ConstraintInterpreter implements ConstraintTranslator<Optional<ValidationError>> {
```

Comment:
As discussed, among the triad `ConstraintTranslator`, `ConstraintInterpreter` and `ConstraintComplier` it takes some several seconds to realize which is which. Let's try to rename them so that each would sound straight "in your face" and require minimal energy for identification.